A document type for error messages

Octachron opened the pull request:
Currently when the compiler needs to build a fragment of an error message while playing nice with `Format`, the best option is to construct a `Format.formatter -> unit` closure. This construction is used in many part of the compilers.
However, it has proven to be quite brittle due to the simultaneous uses of many global states when printing error messages, see #13099 for a recent example (which was triggered by a printing closure run in a different environment that the one it was built in).

This PR proposes to remove this brittleness by a introducing a `document` type which represents a list of formatting instructions 
```ocaml
type element =
  | Text of string
  | Open_box of { kind: box_type ; indent:int }
  | Close_box
  | Open_tag of Format.stag
  | Close_tag
  |  ...
```
to be sent to an actual formatting engine at a later point. Since this `document` type contains only data and no closures (except for an escape hatch), we are guaranteed that printing this type does not create side-effects without losing  any of the formatting abilities of `Format`. In particular, this document type can be serialized while preserving all the break hints, formatting boxes and tags (which semantics could then be reimplemented by developer tools). 

In term of implementation, this PR adds a new `Format_doc` module which provides:
* format string interpreters that output this new `document` type
* printing functions that mirrors the printing functions in the `Format` module
* a conversion function from `Format_doc` printer to `Format` printer
* a "deprecated" conversion function in the other direction which uses a temporary escape hatches
* A  submodule for building `document` in an immutable way
 
As a consequence the `Format_doc` module is source-compatible with the `Format` module, and for most of the compiler modules replacing `open Format` by an `open Format_doc` is enough to switch to the new kinds of printers.

Octachron committed:
Format_doc: a new data type for composing user-facing text messages

Octachron committed:
error messages: replace Format by Format_doc

Octachron committed:
Disable name disambiguation for paths without namespace

Octachron committed:
format_doc: less delayed printers in printtyp

gasche commented on utils/format_doc.ml:
```diff
@@ -0,0 +1,476 @@
+  | Break of { fits : string * int * string as 'a; breaks : 'a }
+  | Flush of { newline:bool }
+  | Newline
+  | If_newline
```
I was surprised by `If_newline` which is an odd construct (very imperative). @Octachron tells me that it is not actually used in the compiler so we could get rid of it if we wanted -- same with `With_size` which is also weird. I don't know if we should rather follow the oddities of Format to the letter, or remove the stuff we don't like and don't need.

jberdine replied:
IIRC `If_newline` predates the form of `Break` here, and I wouldn't be surprised if `If_newline` is not really needed.

If you ever want to include static unicode text/symbols in diagnostics, then I expect `With_size` could be useful. If the unicode you want to emit is known ahead of time, the complicated question of computing its width can be worked around by specifying the size. `Format` itself works well in this way.

gasche commented on utils/format_doc.ml:
```diff
@@ -0,0 +1,476 @@
+let empty : doc = { rev = [] }
+
+let to_list doc = List.rev doc.rev
+let add x doc = { rev = x :: doc.rev }
```
I would find the parameter order `add doc elem` more natural than `add elem doc`, because the element is printed after the rest of the document, not before.

gasche commented on utils/format_doc.mli:
```diff
@@ -0,0 +1,281 @@
+
+end
+
+(** {1 Compatibility API} *)
```
We could have the functions that are not so nice and mostly for compatibility in a `Compat` submodule. But I am okay with the current approach as well, the functions in fact tend to make sense already.

gasche commented on testsuite/tests/typing-poly/error_messages.ml:
```diff
@@ -58,11 +58,11 @@ Lines 5-7, characters 10-5:
-       The universal variable "'a" would escape its scope
+       The method "f" has type "'a -> int", but the expected method type was
+       "'a0. 'a0 -> int"
+       The universal variable "'a0" would escape its scope
```
Note to self: I liked seeing `'a` and `´b` better than `'a` and `'a0` (personally I am fine with any order of occurrence, but I like the variables to be easy to distinguish). But maybe in some other examples reusing the same radical helps readability?

gasche commented on typing/printtyp.ml:
```diff
@@ -2592,25 +2585,32 @@ let error trace_format mode subst env tr txt1 ppf txt2 ty_expect_explanation =
   match tr with
   | [] -> assert false
-  | elt :: tr ->
+  | elt :: tr as full_trace ->
```
(I would find `(elt :: tr) as full_trace` more readable.)

gasche commented on typing/printtyp.ml:
```diff
@@ -2280,13 +2280,14 @@ let prepare_trace f tr =
 
 (** Keep elements that are [Diff _ ] and take the decision
     for the last element, require a prepared trace *)
```
I wish the documentation string would be clear that the last element is *not* included in the first component of the returned value. (I would have guessed otherwise.)

gasche reviewed (commented):
This is nice and overall quite reasonable. I asked a few questions. I don't understand the very fine details of the last commit, which I think is the most subtle one. (The rest I believe are correct.)

The fact that the testsuite passes mostly unchanged is a very good test for the PR, so I am confident in its overall correctness.

Octachron committed:
update Changes

Julow commented on utils/format_doc.mli:
```diff
@@ -0,0 +1,290 @@
+
+(** {1 Compatibility API} *)
+
+(** The functions and types below provides source compatibility with format
```
Opening this module adds the document types, `fold`, etc.. into the environment. What do you think of moving the compatibility API into a `Compat` submodule ?
Also, highlighting that it's at the same level as the `Core` API.

Octachron replied:
One my design goal was to make the replacement of `Format` by `Format_doc` as invisible as possible.
Typically, the `Core` submodule is only exposed because I though it would be a shame to have a simple immutable API available and hide it away. I am not sure I would encourage people to switch to this API.

From that perspective, I think it is better for now to keep the most used features (the compatibility functions) at the toplevel of the module. But I could hide the core definitions inside the `Core` submodule?

Octachron replied:
I have moved the type definitions and core functions inside the `Core` submodule, and renamed it to `Doc`. Like this, an `open Format_doc` only opens the compatibility API and the `Doc` submodule.

Julow commented on utils/format_doc.ml:
```diff
@@ -0,0 +1,476 @@
+  | Flush {newline=false} -> Format.pp_print_flush ppf ()
+  | Newline -> Format.pp_force_newline ppf ()
+  | If_newline -> Format.pp_print_if_newline ppf ()
+  | With_size _ ->  ()
```
Some functions like `pp_print_substring_as`, `with_size` and `if_newline` are not used. Is offering an alternative API to Format a goal ?
Otherwise, why not keep it smaller to help maintenance ?

You mentioned localization in https://github.com/ocaml/ocaml/pull/12182#issuecomment-2042180005. Could it be built on top of the this API ? Making it as small as needed could help adapting it later.

Octachron replied:
My first goal was full source compatibility with `Format`, since this means compiler contributors don't need to know two APIs.

(And if we were to move away from `Format` compatibility, I would rather adopt `Fmt` naming convention)

For basic localization, a simple way to add support for compiler variant would be to add hooks for the format string interpreters since we only want to localize static texts.

Octachron committed:
Format_doc: move definitions inside Doc submodule

This ensures that only the compatibility API is available after an
`open Format_doc` used to replace an `open Format`.

gasche commented on parsing/location.mli:
```diff
@@ -169,11 +169,16 @@ val show_filename: string -> string
+  val print_loc: formatter -> t -> unit
+  val print_locs: formatter -> t list -> unit
+  val separate_new_message: formatter -> unit -> unit
+end
```
I am nervous about potential usage of `Location` from compiler-libs, in particular using `Location.print_loc`. To guarantee compatibility I would rather leave the existing functions unchanged and add a `Doc` submodule with the new stuff.

Octachron replied:
This sounds sensible, I moved the new printer inside a `Doc` submodule, and added a `quoted_filename` function along the way.

gasche commented on utils/format_doc.ml:
```diff
@@ -0,0 +1,477 @@
+  let rec list ?(sep=Fun.id) elt l doc = match l with
+    | [] -> doc
+    | [a] -> elt a doc
+    | a :: (_ :: _ as q) ->
```
Personally I find `((_ :: _) as q)` more readable -- basically I don't like the current associativity of `as` which is unlike other infix operators. Feel free to ignore.

Octachron replied:
Changed, since I don't have strong opinions on that subject.

gasche reviewed (commented):
There is some code duplication between utils/format_doc.ml and stdlib/format.ml, for example the printf-interpretation functions (`output_acc` and friends), and some higher-level printers such as `pp_print_text`, which decompose into calls into more basic `pp_print_*` functions.

I wonder if it would be reasonable to reduce this duplication by introducing a functor that builds this higher-level structure over the `pp_print_*` functions. One instance would be in stdlib/format.ml, the other in utils/format_doc.ml. One pain point is that the functor should be in the stdlib itself, which is slightly annoying -- camlinternalFormatter.ml?

Octachron committed:
review: move new Location printers inside Location.Doc

and let the compatibility function at the top of the Location modules
to improve backward compatibility for compilerlibs users.

* add a `Doc.quoted_filename` function too.

Octachron commented:
There certainly some part that could be factorized using a `Format-like` functor builder, but it doesn't seem that straightforward to find a home for this functor inside the standard library. Moreover, there are also some interesting design choices to discuss in term of functor complexity versus flexibility.

Thus, I would rather report the standard library bikeshedding to another PR.

Octachron commented:
I have updated the PR after some noise clean-up, I plan to merge it tomorrow.

gasche commented:
This PR breaks the build of Coq, which contains the following code:

https://github.com/coq/coq/blob/729af0da7a829d08223f54f6249235c888275c80/topbin/coqtop_byte_bin.ml#L11-L17

The problem is that `Symtable.report_error` has changed type from `formatter -> error -> unit` to an incompatible type `t Format_doc.printer`.

@Octachron, do you have an idea of how the user code could be fixed, preferably in a backward-compatible way?

Octachron commented:
Replacing `Format.asprintf` by `Format_doc.asprintf` should be enough. A backward compatible fix would be to select either `Format` or `Format_doc` in function of the OCaml compiler library version.

gasche commented:
Okay, I went along and wrote a commit that adds conditional-compilation support to Coq for this one function, to make it work correctly both before and after 5.3 (I haven't submitted it as a PR for reasons that are clarified below):

https://github.com/coq/coq/compare/master...gasche:coq:compat-symtable-error-handler

And then I realized that there are probably more projects affected by this sort of breakage.

For example if I run a [github search](https://github.com/search?q=language%3AOCaml+%22report_error%22&type=code) with `language:OCaml "report_error"`, on the first page out of 5 I see two distinct projects that use `report_error` functions coming from the compiler codebase:

https://github.com/BinaryAnalysisPlatform/bap/blob/95e81738c440fbc928a627e4b5ab3cccfded66e2/src/baptop.ml#L7-L10

https://github.com/rescript-lang/rescript-compiler/blob/7cdb66d5ad09f4ea49af6f3a6a8aa0c08adefab7/jscomp/ml/env.ml#L753-L756

I haven't gone through the rest of the list, I'm sure there are other affected projects.

I am in favor of document types for error messages, but this is too must breakage. Could we rework this to keep the `report_error` functions with their previous type for backwards-compatibility, and introduces `error_doc` variants with the new incompatible type?

Octachron commented:
Rescript is a divergent fork of OCaml 4.06 which is not affected by the issue.
The BAP case could be fixed by using `Location.errorf` to be backward compatible with all OCaml version.
The only other example in the 5 page github search seems to be `ocamleditor`.

Thus we have two and half breakages? We could keep the deprecated version, and I would probably submit a PR do so this week; however it is unclear to me if it would not be faster to fix the handful of broken packages.

jmid commented:
FTR, I encountered `refl.4.1.0` breaking on 5.3.0 due to this change to `Location.raise_errorf`: https://github.com/ocaml/opam-repository/pull/28364
